src/ess/polarization/base.py

```diff
@@ -0,0 +1,200 @@
+RawEventData = NewType('RawEventData', sc.DataArray)
+
+
+class He3Transmission(sl.Scope[Cell, sc.DataArray], sc.DataArray):
+    """Spin-, Time-, and wavelength-dependent transmission for a given cell."""
+
+
+class He3CellPressure(sl.Scope[Cell, sc.Variable], sc.Variable):
+    """Pressure for a given cell."""
+
+
+class He3CellLength(sl.Scope[Cell, sc.Variable], sc.Variable):
+    """Length for a given cell."""
+
+
+class He3FillingTime(sl.Scope[Cell, sc.Variable], sc.Variable):
+    """Filling wall-clock time for a given cell."""
+
+
+class He3Opacity(sl.Scope[Cell, sc.DataArray], sc.DataArray):
+    """Wavelength-dependent opacity for a given cell."""
+
+
+class DirectBeamTimeIntervals(sl.ScopeTwoParams[Cell, Spin, sc.Variable], sc.Variable):
+    """
+    Wall-clock time intervals for a given cell.
+
+    This defines the time intervals that correspond the direct beam measurements. This
+    is used to extract the direct beam data from the total event data.
+    """
+
+
+class He3DirectBeam(sl.ScopeTwoParams[Cell, Spin, sc.DataArray], sc.DataArray):
+    """
+    Direct beam data for a given cell and spin state.
+
+    TODO How is this defined? It is a processed version of the raw direct beam event
+    data.
+    """
+
+
+class He3InitialAtomicPolarization(sl.Scope[Cell, sc.DataArray], sc.DataArray):
+    """
+    Initial atomic polarization for a given cell.
+
+    This is computed from the direct beam data with an unpolarized cell.
+    """
+
+
+class SampleData(
+    sl.ScopeTwoParams[PolarizerSpin, AnalyzerSpin, sc.DataArray], sc.DataArray
+):
+    """
+    Uncorrected sample data.
+
+    TODO How exactly is this defined? Which normal (non-polarization) corrections
+    have been applied already?
+    """
+
+
+class PolarizationCorrectedSampleData(
+    sl.ScopeTwoParams[PolarizerSpin, AnalyzerSpin, sc.DataArray], sc.DataArray
+):
+    """Polarization-corrected sample data."""
+
+
+class SampleTimeIntervals(
+    sl.ScopeTwoParams[PolarizerSpin, AnalyzerSpin, sc.Variable], sc.Variable
+):
+    """
+    Wall-clock time intervals for a given sample.
+
+    This defines the time intervals that correspond the sample measurements. This is
+    used to extract the sample data from the total event data.
+    """
+
+
+def he3_direct_beam(
+    event_data: RawEventData,
+    time_intervals: DirectBeamTimeIntervals[Cell, Spin],
+    wavelength: WavelengthBins,
+) -> He3DirectBeam[Cell, Spin]:
+    """
+    Returns the direct beam data for a given cell and spin state.
+
+    We have a sequence of direct beam measurements, e.g., defined by a list of wall
+    clock time intervals.
+    """
+    # compute wavelength
+    # return event data, dims=(interval,wavelength)
+
+
+def he3_opacity(
+    pressure: He3CellPressure[Cell],
+    cell_length: He3CellLength[Cell],
+    wavelength: WavelengthBins,
+) -> He3Opacity[Cell]:
+    return He3Opacity[Cell]()
+
+
+def he3_initial_atomic_polarization(
+    direct_beam: He3DirectBeam[Cell, Unpolarized],
```

Comment:
Ok, after reviewing the functions again (uploaded in issues and send by email) here following changes are required:
- change naming"unpolarized" to "unpolarized beam" (as it needs to be differentiated from "unpolarized cell"
- we need two different unpolarized beam measuremetns here: with the polarized 3He cell in, and with 3He cell out
---> change "He3DirectBeam[Cell, Unpolarized]" to:
He3DirectBeam[Cell, Unpolarized beam, polarized 3He cell in]
&
He3DirectBeam[Cell, Unpolarized beam, polarized 3He cell out]

Reply:
> we need two different unpolarized beam measuremetns here

Will they both use the same function, or do we need to split it?

Reply:
> ---> change "He3DirectBeam[Cell, Unpolarized]" to:
He3DirectBeam[Cell, Unpolarized beam, polarized 3He cell in]
&
He3DirectBeam[Cell, Unpolarized beam, polarized 3He cell out]

- I thought this would be indicated by the `Cell`, which can be `Polarizer` and `Analyzer`?
- How can the beam be unpolarized of a polarized He3 cell is in?

Reply:
Maybe its easier to have a list of following states:
[Cell] = [polarizer, analyzer]
[Spin] =  [Up, Down]
[3He polarization state] = [polarized, unpolarized]
[Cell position] = [cell in,cell out]

--> and then add here to change it to:
He3DirectBeam[Cell, Unpolarized beam, Cell position]

Reply:
Can you explicitly list all the different direct-beam measurements that will be performed?

Reply:
> > we need two different unpolarized beam measuremetns here
> 
> Will they both use the same function, or do we need to split it?

Same function, to compute the initial atomic polarization

Reply:
> > ---> change "He3DirectBeam[Cell, Unpolarized]" to:
> > He3DirectBeam[Cell, Unpolarized beam, polarized 3He cell in]
> > &
> > He3DirectBeam[Cell, Unpolarized beam, polarized 3He cell out]
> 
>     * I thought this would be indicated by the `Cell`, which can be `Polarizer` and `Analyzer`?
--> I thought 'Cell' only defines which cell? the new parameter set [cell position] = [cell in, cell out] will determine if its in the beam or not, i.e., if we measure transmission through the cell or if we take it out before as reference measurement

>     * How can the beam be unpolarized of a polarized He3 cell is in?
-->this probably needs a more precise naming, namely "unpolarized incident beam" (i.e., if its the analyzer, then the polarizer is out. For Dream that will be always the case for the 3He measurements as there are just two 3He cells, but e.g. for SKADI with a supermirror analyzer that will be important to be differentiated if the polarizer is in or out, i.e., if the neutron beam coming to the analyzer is polarized or not...)

Reply:
> * a direct-beam measurement _for the polarizer_?
--> this is a reference measurement for the polarizer. basically taking the quotien (Measurement polarizer in)/(Measurement polarizer out)

Reply:
> Can you explicitly list all the different direct-beam measurements that will be performed?

I will make a nice overview, get it reviewed by Hal, and post it here.